### PR TITLE
Account_notes plugin, bandwidth for custom_json #924

### PR DESCRIPTION
**Pull request by maslenitsa93**
#924 

# Checklists

Use testnet.

### Main flow

1. Create an account and create the few fields:
```
./cli_wallet --server-rpc-endpoint=ws://127.0.0.1:8091
set_password qwer
unlock qwer
import_key 5JVFFWRLwz6JoP9kguuRFfytToGU6cLgBVTL9t6NB3D3BQLbUBS
create_account cyberfounder test "{}" "30.000 GOLOS" true
transfer_to_vesting cyberfounder test "30.000 GOLOS" true
begin_builder_transaction
add_operation_to_builder_transaction 0 ["custom_json", {"required_auths":["test"],"id":"account_notes","json":"[\"set_value\", {\"account\":\"test\",\"key\":\"lorem\",\"value\":\"ipsum\"}]"}]
sign_builder_transaction 0 true
begin_builder_transaction
add_operation_to_builder_transaction 1 ["custom_json", {"required_auths":["test"],"id":"account_notes","json":"[\"set_value\", {\"account\":\"test\",\"key\":\"abcdef\",\"value\":\"123456\"}]"}]
sign_builder_transaction 1 true
```

2. Run following script in browser to get values:
```
<script src="node_modules/golos-js/dist/golos.min.js"></script>
Hello world
<script>
golos.config.set('websocket', 'ws://127.0.0.1:8091');
golos.api.send("account_notes", {"method": "get_value", "params":["test","lorem"]},
    function (err, result) {
document.body.innerHTML += 'lorem: <pre>' + JSON.stringify(result, null, 2) + "</pre>";
});
golos.api.send("account_notes", {"method": "get_value", "params":["test","abcdef"]},
    function (err, result) {
document.body.innerHTML += 'abcdef: <pre>' + JSON.stringify(result, null, 2) + "</pre>";
});
</script>
```
Result:
lorem - ipsum
abcdef - 123456

3. Edit one of values with cli_wallet:
```
begin_builder_transaction
add_operation_to_builder_transaction 2 ["custom_json", {"required_auths":["test"],"id":"account_notes","json":"[\"set_value\", {\"account\":\"test\",\"key\":\"abcdef\",\"value\":\"7890\"}]"}]
sign_builder_transaction 2 true
```

4. Run above browser script again. Result:
lorem - ipsum
abcdef - 7890

### Length limits

```
begin_builder_transaction
add_operation_to_builder_transaction 3 ["custom_json", {"required_auths":["test"],"id":"account_notes","json":"[\"set_value\", {\"account\":\"test\",\"key\":\"123456789012345678901\",\"value\":\"normal_value_big_key\"}]"}]
sign_builder_transaction 3 true
```
Result: Key is too long.

```
begin_builder_transaction
add_operation_to_builder_transaction 4 ["custom_json", {"required_auths":["test"],"id":"account_notes","json":"[\"set_value\", {\"account\":\"test\",\"key\":\"normal_key_big_value\",\"value\":\"123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123\"}]"}]
sign_builder_transaction 4 true
```
Result: Value is too long.

### Case with non-existant note

Run following script in browser:
```
<script src="node_modules/golos-js/dist/golos.min.js"></script>
Hello world
<script>
golos.config.set('websocket', 'ws://127.0.0.1:8091');
golos.api.send("account_notes", {"method": "get_value", "params":["test","not_existant"]},
    function (err, result) {
document.body.innerHTML += 'not_existant: <pre>' + JSON.stringify(result, null, 2) + "</pre>";
});
</script>
```
Result: ""

### Note count limit. Deleting notes. 

1. On empty node, run this:
```
./cli_wallet --server-rpc-endpoint=ws://127.0.0.1:8091
set_password qwer
unlock qwer
import_key 5JVFFWRLwz6JoP9kguuRFfytToGU6cLgBVTL9t6NB3D3BQLbUBS
create_account cyberfounder test "{}" "30.000 GOLOS" true
transfer_to_vesting cyberfounder test "30.000 GOLOS" true
create_account cyberfounder test2 "{}" "30.000 GOLOS" true
transfer_to_vesting cyberfounder test2 "30.000 GOLOS" true
begin_builder_transaction
add_operation_to_builder_transaction 0 ["custom_json", {"required_auths":["test2"],"id":"account_notes","json":"[\"set_value\", {\"account\":\"test2\",\"key\":\"outside_of_test\",\"value\":\"123\"}]"}]
sign_builder_transaction 0 true
begin_builder_transaction
add_operation_to_builder_transaction 1 ["custom_json", {"required_auths":["test"],"id":"account_notes","json":"[\"set_value\", {\"account\":\"test\",\"key\":\"lorem\",\"value\":\"ipsum\"}]"}]
sign_builder_transaction 1 true
begin_builder_transaction
add_operation_to_builder_transaction 2 ["custom_json", {"required_auths":["test"],"id":"account_notes","json":"[\"set_value\", {\"account\":\"test\",\"key\":\"lorem2\",\"value\":\"ipsum\"}]"}]
sign_builder_transaction 2 true
begin_builder_transaction
add_operation_to_builder_transaction 3 ["custom_json", {"required_auths":["test"],"id":"account_notes","json":"[\"set_value\", {\"account\":\"test\",\"key\":\"lorem3\",\"value\":\"ipsum\"}]"}]
sign_builder_transaction 3 true
begin_builder_transaction
add_operation_to_builder_transaction 4 ["custom_json", {"required_auths":["test"],"id":"account_notes","json":"[\"set_value\", {\"account\":\"test\",\"key\":\"lorem4\",\"value\":\"ipsum\"}]"}]
sign_builder_transaction 4 true
begin_builder_transaction
add_operation_to_builder_transaction 5 ["custom_json", {"required_auths":["test"],"id":"account_notes","json":"[\"set_value\", {\"account\":\"test\",\"key\":\"lorem5\",\"value\":\"ipsum\"}]"}]
sign_builder_transaction 5 true
begin_builder_transaction
add_operation_to_builder_transaction 6 ["custom_json", {"required_auths":["test"],"id":"account_notes","json":"[\"set_value\", {\"account\":\"test\",\"key\":\"lorem6\",\"value\":\"ipsum\"}]"}]
sign_builder_transaction 6 true
begin_builder_transaction
add_operation_to_builder_transaction 7 ["custom_json", {"required_auths":["test"],"id":"account_notes","json":"[\"set_value\", {\"account\":\"test\",\"key\":\"lorem7\",\"value\":\"ipsum\"}]"}]
sign_builder_transaction 7 true
begin_builder_transaction
add_operation_to_builder_transaction 8 ["custom_json", {"required_auths":["test"],"id":"account_notes","json":"[\"set_value\", {\"account\":\"test\",\"key\":\"lorem8\",\"value\":\"ipsum\"}]"}]
sign_builder_transaction 8 true
begin_builder_transaction
add_operation_to_builder_transaction 9 ["custom_json", {"required_auths":["test"],"id":"account_notes","json":"[\"set_value\", {\"account\":\"test\",\"key\":\"lorem9\",\"value\":\"ipsum\"}]"}]
sign_builder_transaction 9 true
begin_builder_transaction
add_operation_to_builder_transaction 10 ["custom_json", {"required_auths":["test"],"id":"account_notes","json":"[\"set_value\", {\"account\":\"test\",\"key\":\"lorem10\",\"value\":\"ipsum\"}]"}]
sign_builder_transaction 10 true
```
Result: OK

2. Now, add one more note: 
```
begin_builder_transaction
add_operation_to_builder_transaction 11 ["custom_json", {"required_auths":["test"],"id":"account_notes","json":"[\"set_value\", {\"account\":\"test\",\"key\":\"lorem11\",\"value\":\"ipsum\"}]"}]
sign_builder_transaction 11 true
```
Result: Note count limit exceeded.

3. Delete one note:
```
begin_builder_transaction
add_operation_to_builder_transaction 12 ["custom_json", {"required_auths":["test"],"id":"account_notes","json":"[\"set_value\", {\"account\":\"test\",\"key\":\"lorem10\",\"value\":\"\"}]"}]
sign_builder_transaction 12 true
```

4. Try add again:
```
begin_builder_transaction
add_operation_to_builder_transaction 13 ["custom_json", {"required_auths":["test"],"id":"account_notes","json":"[\"set_value\", {\"account\":\"test\",\"key\":\"lorem11\",\"value\":\"ipsum\"}]"}]
sign_builder_transaction 13 true
```
Result: OK.

### Account white and black lists

1. Check white list.

```
an-tracked-accounts = ["test2"]
```

```
./cli_wallet --server-rpc-endpoint=ws://127.0.0.1:8091
set_password qwer
unlock qwer
import_key 5JVFFWRLwz6JoP9kguuRFfytToGU6cLgBVTL9t6NB3D3BQLbUBS
create_account cyberfounder test "{}" "30.000 GOLOS" true
transfer_to_vesting cyberfounder test "30.000 GOLOS" true
create_account cyberfounder test2 "{}" "30.000 GOLOS" true
transfer_to_vesting cyberfounder test2 "30.000 GOLOS" true
begin_builder_transaction
add_operation_to_builder_transaction 0 ["custom_json", {"required_auths":["test2"],"id":"account_notes","json":"[\"set_value\", {\"account\":\"test2\",\"key\":\"abcdef\",\"value\":\"123456\"}]"}]
sign_builder_transaction 0 true
begin_builder_transaction
add_operation_to_builder_transaction 1 ["custom_json", {"required_auths":["test"],"id":"account_notes","json":"[\"set_value\", {\"account\":\"test\",\"key\":\"lorem\",\"value\":\"ipsum\"}]"}]
sign_builder_transaction 1 true
```

Result: 
test2 - OK
test - failure

2. Check black list.

```
an-untracked-accounts = ["test"]
```

Same script.

Result: 
test2 - OK
test - failure

### get_settings API method

Run following script in browser:
```
<script src="node_modules/golos-js/dist/golos.min.js"></script>
Hello world
<script>
golos.config.set('websocket', 'ws://127.0.0.1:8091');
golos.api.send("account_notes", {"method": "get_settings", "params":[]},
    function (err, result) {
document.body.innerHTML += 'not_existant: <pre>' + JSON.stringify(result, null, 2) + "</pre>";
});
</script>
```
Result example:
```
{
"max_key_length": 20,
"max_value_length": 512,
"max_note_count": 10,
"tracked_accounts": [
"alice",
"bob"
],
"untracked_accounts": []
}
```

### CODE REVIEW

File: plugins/account_notes/account_notes_operations.cpp [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+namespace golos { namespace plugins { namespace account_notes {
+
+using golos::protocol::is_valid_account_name;
+
+static inline void validate_account_name(const string& name) {
+    GOLOS_CHECK_VALUE(is_valid_account_name(name), "Account name ${name} is invalid", ("name", name));
+}
+
+void set_value_operation::validate() const {
+    GOLOS_CHECK_PARAM_ACCOUNT(account);
```
[the remark this answers is not in the document]
Comment (maslenitsa93) [reply plugins/account_notes/account_notes_operations.cpp]: @afalaleev Here, in validate(), should be checking for 128 and 4096. But how it correlates with configurable settings in plugin?

[the remark this answers is not in the document]
Comment (afalaleev) [reply plugins/account_notes/account_notes_operations.cpp]: 128 and 4096 for chain and it should be checked in validate()

Settings in plugin is for node, and they should be checked in evaluator.

File: plugins/account_notes/account_notes_evaluators.cpp [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            return;
+        }
+
+        if (notes_itr != notes_idx.end()) { // Edit case
+            db().modify(*notes_itr, [&](account_note_object& n) {
+                from_string(n.value, o.value);
+            });
+        } else { // Create case
+            const auto& count_idx = db().get_index<account_note_index, by_account>();
+            auto notes = count_idx.count(o.account);
```
Comment (afalaleev) [inline plugins/account_notes/account_notes_evaluators.cpp]: You should create an additional index with statistic for key-values per user.

File: plugins/account_notes/account_notes_evaluators.cpp [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+#include <golos/plugins/account_notes/account_notes_evaluators.hpp>
+#include <golos/chain/account_object.hpp>
+
+namespace golos { namespace plugins { namespace account_notes {
+
+using golos::chain::from_string;
+
+bool set_value_evaluator::is_tracked_account(const account_name_type& account) {
+    if (!settings_->tracked_accounts.empty()
+        && std::find(settings_->tracked_accounts.begin(), settings_->tracked_accounts.end(),
```
Comment (afalaleev) [inline plugins/account_notes/account_notes_evaluators.cpp]: flat_set has method `find()` and `count`

File: plugins/account_notes/account_notes_evaluators.cpp [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    return true;
+}
+
+void set_value_evaluator::do_apply(const set_value_operation& o) {
+    try {
+        if (!is_tracked_account(o.account)) {
+            return;
+        };
+
+        db().get_account(o.account);
```
Comment (afalaleev) [inline plugins/account_notes/account_notes_evaluators.cpp]: You already check account on exist in `get_required_active_authorities` of operation.

File: plugins/account_notes/account_notes_evaluators.cpp (hunk 2 above)
Comment (afalaleev) [reply plugins/account_notes/account_notes_evaluators.cpp]: It is a bad idea to iterate via items.

File: libraries/chain/database.cpp [hunk 5]
```diff
@@ -796,8 +796,8 @@ namespace golos { namespace chain {
             adjust_supply(-fee);
         }
 
-        bool database::update_account_bandwidth(const account_object &a, uint32_t trx_size, const bandwidth_type type) {
-            const auto &props = get_dynamic_global_properties();
+        bool database::update_account_bandwidth(const dynamic_global_property_object& props, const account_object &a, uint32_t trx_size,
+                const bandwidth_type type) {
```
Comment (afalaleev) [inline libraries/chain/database.cpp]: Why 2 lines?

File: plugins/account_notes/account_notes_evaluators.cpp [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+#include <golos/plugins/account_notes/account_notes_evaluators.hpp>
+#include <golos/chain/account_object.hpp>
+
+namespace golos { namespace plugins { namespace account_notes {
+
+using golos::chain::from_string;
+
+bool set_value_evaluator::is_tracked_account(const account_name_type& account) {
+    if (!settings_->tracked_accounts.empty()
+        && settings_->tracked_accounts.find(std::string(account)) == settings_->tracked_accounts.end()) {
```
Comment (afalaleev) [inline plugins/account_notes/account_notes_evaluators.cpp]: `empty()` is not needed, `find()` will return `end()` in this case.

File: plugins/account_notes/account_notes_operations.cpp [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    GOLOS_CHECK_PARAM(key, GOLOS_CHECK_VALUE_LEGE(key.length(), 0, 128));
```
Comment (afalaleev) [inline plugins/account_notes/account_notes_operations.cpp]: Here should be: `GOLOS_CHECK_PARAM(key, GOLOS_CHECK_VALUE_LEGE(key.length(), 1, 128));`

**Pull request merged**